"Get Started" welcome widget

martinfalk opened the pull request:
* Get Started screen provides an overview over recently used workspaces and available examples next to the latest changes.
    + Shortcuts <1> to <0> will select the corresponding recently used workspace
    + <ctrl>+<:digit:> will immediately load the workspace
    + double click on any entry in the left pane will open it (holding <ctrl> will open example workspaces and keep their filename similar to the file menu)
* When saving a workspace, all visible canvases and a snapshot of the workspace will be serialized as well (`WorkspacePreview`)

<img width="600px" src="https://user-images.githubusercontent.com/9251300/51200051-afafc980-18f8-11e9-86b7-239a8dad84d5.png"/>

martinfalk committed:
Qt: serialize workspace image

petersteneteg committed:
Core: Added a save mode to the workspace manager

martinfalk committed:
Qt: welcome widget

martinfalk committed:
Core: fix WorkspaceManager
* processor network was no longer saved

martinfalk committed:
Resources: icons and stylesheet for welcome widget

martinfalk committed:
Qt: welcome widget
* proper layout
* populated file list from recent workspaces and examples

martinfalk committed:
Qt: integration of welcome widget in Qt application

martinfalk committed:
Core: fix WorkspaceSaveMode

martinfalk committed:
Qt,QtWidgets: functionality for workspace preview images

martinfalk committed:
Qt: Get Started screen
* tabbed widgets for network editor and welcome widget
* refactoring of WelcomeWidget and FileTreeWidget
* minor stylesheet adjustments

martinfalk committed:
Qt: WelcomeEditor
* enter key loads the currently selected workspace

martinfalk committed:
Qt: compile fixes

martinfalk committed:
Qt: fix drag&drop of workspaces
* dropping workspace files onto the network editor fixed

martinfalk committed:
QtWidgets: allow renaming ButtonProperty, closes #355
* added renaming of displayName to context menu of ButtonProperty

martinfalk committed:
Qt: WelcomeWidget fixes

ResearchDaniel commented on include/inviwo/qt/editor/inviwomainwindow.h:
```diff
@@ -75,7 +77,7 @@ class IVW_QTEDITOR_API InviwoMainWindow : public QMainWindow, public NetworkEdit
 
     void openLastWorkspace(std::string workspace = "");
-    void openWorkspace(QString workspaceFileName);
+    bool openWorkspace(QString workspaceFileName);
```
Returns true if?

petersteneteg replied:
Might rename askOpenWorkspace?

martinfalk replied:
added comment, returns false in case the workspace could not be opened.

ResearchDaniel commented on include/inviwo/qt/editor/inviwomainwindow.h:
```diff
@@ -99,8 +101,8 @@ class IVW_QTEDITOR_API InviwoMainWindow : public QMainWindow, public NetworkEdit
 
-    void newWorkspace();
-    void openWorkspace();
+    bool newWorkspace();
```
Returns true if?

ResearchDaniel commented on include/inviwo/qt/editor/inviwomainwindow.h:
```diff
@@ -99,8 +101,8 @@ class IVW_QTEDITOR_API InviwoMainWindow : public QMainWindow, public NetworkEdit
-    void newWorkspace();
-    void openWorkspace();
+    bool newWorkspace();
+    bool openWorkspace();
```
Returns true if?

martinfalk replied:
added comments

ResearchDaniel commented on include/inviwo/qt/editor/inviwomainwindow.h:
```diff
@@ -113,18 +115,23 @@ class IVW_QTEDITOR_API InviwoMainWindow : public QMainWindow, public NetworkEdit
     void visibilityModeChangedInSettings();
 
-    void openWorkspace(QString workspaceFileName, bool exampleWorkspace);
+    bool openWorkspace(QString workspaceFileName, bool exampleWorkspace);
```
Returns true if?
We might change "exampleWorkspace" to "readOnly" since that it what we mean?

martinfalk replied:
not really "readonly". If `exampleWorkspace` is true, the workspace file name is not set. Thereby preventing the user from accidentally overwriting the example.

ResearchDaniel commented on include/inviwo/qt/editor/workspacepreview.h:
```diff
@@ -0,0 +1,79 @@
+public:
+    using ImageVector = std::vector<std::pair<std::string, QImage>>;
+
+    struct Item : public Serializable {
```
Item->Preview?

martinfalk replied:
will be renamed to `Base64Image`

ResearchDaniel reviewed (approved):
Nice! 
Some minor comments

petersteneteg commented on include/inviwo/qt/editor/inviwomainwindow.h:
```diff
@@ -113,18 +115,23 @@ class IVW_QTEDITOR_API InviwoMainWindow : public QMainWindow, public NetworkEdit
     virtual void dropEvent(QDropEvent* event) override;
 
 private:
+    friend WelcomeWidget;
```
Remove?

martinfalk replied:
either `WelcomeWidget` is a friend of InviwoMainWindow, or we need to make 
`bool openWorkspace(QString workspaceFileName, bool exampleWorkspace);`
and `QStringList getRecentWorkspaceList() const;` public.

martinfalk replied:
`openWorkspace(QString, bool)`  should not necessarily be exposed in my opinion.

petersteneteg replied:
make them public, thats way better then friends....

petersteneteg commented on src/qt/editor/filetreewidget.cpp:
```diff
@@ -0,0 +1,324 @@
+FileTreeWidget::FileTreeWidget(InviwoApplication *app, QWidget *parent)
+    : QTreeWidget{parent}, inviwoApp_(app), fileIcon_{":/inviwo/inviwo_light.png"} {
+    if (!inviwoApp_) {
+        inviwoApp_ = InviwoApplication::getPtr();
```
Why?

petersteneteg commented on src/qt/editor/inviwomainwindow.cpp:
```diff
@@ -499,7 +523,7 @@ void InviwoMainWindow::addActions() {
     }
 
-#ifdef IVW_DEBUG
+#if defined(IVW_DEBUG) || 1
```
Remove

petersteneteg commented on src/qt/editor/inviwomainwindow.cpp:
```diff
@@ -886,6 +929,20 @@ void InviwoMainWindow::saveWorkspace(QString workspaceFileName) {
     fileName = filesystem::cleanupPath(fileName);
 
     try {
+        auto networkImageSerializationHandle = app_->getWorkspaceManager()->onSave(
```
Should now be added once...

martinfalk replied:
Done

petersteneteg commented on src/qt/editor/inviwomainwindow.cpp:
```diff
@@ -957,6 +1014,38 @@ void InviwoMainWindow::saveWorkspaceAsCopy() {
+
+        centralWidget_->tabBar()->setTabButton(0, QTabBar::RightSide, closeBtn);
+    }
+    centralWidget_->setCurrentIndex(centralWidget_->indexOf(welcomeWidget_));
```
Use raise?

martinfalk replied:
will that work within a QTabWidget?

petersteneteg replied:
Think I use that in the in the docked ones at least. for the propertylistwidget...

petersteneteg commented on src/qt/editor/inviwomainwindow.cpp:
```diff
@@ -957,6 +1014,38 @@ void InviwoMainWindow::saveWorkspaceAsCopy() {
+            centralWidget_->removeTab(index);
+        }
+
+        welcomeWidget_->close();
```
Should we close it now or just hide?

martinfalk replied:
either is fine, I guess.

petersteneteg replied:
or close will not delete it? what does close even mean?

martinfalk replied:
close implicitly deleted the widget, will replace it with `hide()` and keep the widget.

petersteneteg commented on src/qt/editor/welcomewidget.cpp:
```diff
@@ -0,0 +1,507 @@
+}
+
+void WelcomeWidget::initChangelog() {
+    changelog_->setHtml(
```
Lol

petersteneteg replied:
Should still do the python thing here...

martinfalk replied:
Agreed. But maybe not as part of this PR?

martinfalk replied:
part of 5fbc0f6

martinfalk committed:
QtWidgets: optional alpha when querying canvas images

martinfalk committed:
Core,Qt: refactored workspace preview into annotations
* WorkspacePreview (Qt) is replaced by WorkspaceAnnotations (Core) and WorkspaceAnnotationsQt
* annotations included in WelcomeWidget

martinfalk committed:
Core,Qt: annotation widget
* allows editing workspace annotations like author, workspace title, description, etc.
* minor fixes

martinfalk commented:
Added new widget for exposing workspace annotations (title, author, description, etc.) and addressed requested changes. 

Workspace annotations and canvas images are serialized along with the workspace file.
(The widget is shown in the lower left corner.)

<img width='600px' src='https://user-images.githubusercontent.com/9251300/51267681-95402380-19be-11e9-9177-9981753837a4.png'/>

martinfalk committed:
QtWidgets: include fix

martinfalk committed:
Qt: compile fix, updated changelog

martinfalk committed:
Qt: include fix

martinfalk committed:
Qt: minor refactoring
* exposed loading of example workspaces and list of recent workspace
* persistent welcome widget

martinfalk committed:
Qt: splitter layout in WelcomeWidget
* adjustable width for changelog
* stylesheet fix for scrollbars in NetworkEditorView

martinfalk committed:
Core,Qt: default workspace author
* added default author entry to system settings

martinfalk committed:
Qt,Tools: html changelog generation
* uses python markdown module to generate a HTML version of CHANGELOG.md
* available as Qt resource
* WelcomeWidget shows html changelog

martinfalk committed:
Changelog: reformatting and clean-up
* clean-up changelog to produce nicer html output

martinfalk commented:
Latest changes:
* PR fixes
* default author in system settings
* new layout in WelcomeWidget using a splitter, which allows to show more/less of the change log
* generating the html changelog is now part of the build process (assuming python and its markdown module are available)

martinfalk committed:
Qt: crash fix
* fixed crash when destructing AnnotationsWidget which is caused by owned properties
